File: Homework-by-python/Homework-Week-4/Hw 4_1 字符串匹配问题.py
def is_valid(s):
    stack = []
    mapping = {
        ')': '(',
        ']': '[',
        '}': '{',
        '>': '<'
    }
    order = ['<', '(', '[', '{']

    for char in s:
        if char in mapping.values():
            if stack and order.index(stack[-1]) < order.index(char):
                return "NO"
            stack.append(char)
        elif char in mapping.keys():  
            if not stack or mapping[char] != stack.pop():
                return "NO"
        else:
            return "NO"
    
    return "YES" if not stack else "NO"

if __name__ == "__main__":
    n = int(input())
    results = []
    for _ in range(n):
        s = input()
        results.append(is_valid(s))
    for res in results:
        print(res)

File: Homework-by-C++/Homework-Week-4/Hw 4_1 字符串匹配问题.cpp
#include <iostream>
#include <stack>
#include <unordered_map>
#include <string>
#include <vector>

using namespace std;

string is_valid(const string& s) {
    stack<char> stack;
    unordered_map<char, char> mapping = {
        {')', '('},
        {']', '['},
        {'}', '{'},
        {'>', '<'}
    };
    string order = "<([{";

    for (char charact : s) {
        if (order.find(charact) != string::npos) {
            if (!stack.empty() && order.find(stack.top()) < order.find(charact))
                return "NO";
            stack.push(charact);
        } else if (mapping.find(charact) != mapping.end()) {
            if (stack.empty() || mapping[charact] != stack.top())
                return "NO";
            stack.pop();
        } else {
            return "NO";
        }
    }

    return stack.empty() ? "YES" : "NO";
}

int main() {
    int n;
    cin >> n;
    vector<string> results;

    for (int i = 0; i < n; ++i) {
        string s;
        cin >> s;
        results.push_back(is_valid(s));
    }

    for (const auto &res : results)
        cout << res << endl;

    return 0;
}
